### PR TITLE
NRG (2.11): Truncate entries without quorum from different `pterm`

### PR DESCRIPTION
**Pull request by MauriceVanVeen**
Reverts the changes made in https://github.com/nats-io/nats-server/pull/5987, but the tests are kept.

Instead opting for a simpler approach:
- removing the `isNew` condition when `pterm` or `pindex` don't match, to ensure consistency even during catchup
- move the `ae.pindex == n.pindex` condition up so `pterm` can be corrected (otherwise it would not be executed)

Signed-off-by: Maurice van Veen <github@mauricevanveen.com>

### CODE REVIEW

File: server/raft.go [hunk 1]
```diff
@@ -3383,35 +3383,37 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
 		n.updateLeadChange(false)
 	}
 
-	if (isNew && ae.pterm != n.pterm) || ae.pindex != n.pindex {
```
Comment (MauriceVanVeen) [inline server/raft.go:3386]: Removing `isNew` makes sure we can also detect differences if we're catching up, otherwise we could wrongly assume our own state is correct.

Specifically covered by the test `TestNRGWALEntryWithoutQuorumMustTruncate/diverged`

File: server/raft.go [hunk 2]
```diff
@@ -3434,16 +3436,6 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
 				return
 			}
 
-			// Check if only our terms do not match here.
-			if ae.pindex == n.pindex {
```
Comment (MauriceVanVeen) [inline server/raft.go:3438]: This condition could never be hit before, due to the condition above:
```go
// Check if this is a lower or equal index than what we were expecting.
if ae.pindex <= n.pindex {
```

Now moved up so the `pterm` can be corrected, otherwise we could spin.

Already covered by the test `TestJetStreamClusterStreamCatchupNoState`. Previously it used a different way to pass, likely truncating a bit too much and eventually converging. Now it only corrects the `pterm` as truncating is not needed.

File: server/raft.go [hunk 3]
```diff
@@ -3485,11 +3477,11 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
 			return
 
 		} else {
-			n.debug("AppendEntry did not match %d %d with %d %d (commit %d)", ae.pterm, ae.pindex, n.pterm, n.pindex, n.commit)
-			if ae.pindex > n.commit {
+			n.debug("AppendEntry did not match %d %d with %d %d", ae.pterm, ae.pindex, n.pterm, n.pindex)
+			if ae.pindex > n.pindex {
```
[the remark this answers is not in the document]
Comment (MauriceVanVeen) [reply server/raft.go]: yep, you're right, removed that if statement

**Review by neilalexander — APPROVED** [review]
LGTM, have spoken to @MauriceVanVeen about some of the changes here and I think they're OK. Should be simpler than the previous approach of starting catchups at `n.commit`.

File: server/raft.go [hunk 4]
```diff
@@ -3383,35 +3383,37 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
 		n.updateLeadChange(false)
 	}
 
-	if (isNew && ae.pterm != n.pterm) || ae.pindex != n.pindex {
-		// Check if this is a lower index than what we were expecting.
-		if ae.pindex < n.pindex {
+	if ae.pterm != n.pterm || ae.pindex != n.pindex {
+		// Check if this is a lower or equal index than what we were expecting.
+		if ae.pindex <= n.pindex {
 			n.debug("AppendEntry detected pindex less than ours: %d:%d vs %d:%d", ae.pterm, ae.pindex, n.pterm, n.pindex)
```
Comment (mprimi) [inline server/raft.go:3389]: (This debug statement is now very slightly incorrect)

File: server/raft.go [hunk 5]
```diff
@@ -3383,35 +3383,37 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
...
+
+			if n.commit > 0 && ae.pindex <= n.commit {
+				// If we have already committed this entry, just mark success.
+				success = true
+			} else if eae, _ := n.loadEntry(ae.pindex); eae == nil {
 				// If terms are equal, and we are not catching up, we have simply already processed this message.
 				// So we will ACK back to the leader. This can happen on server restarts based on timings of snapshots.
 				if ae.pterm == n.pterm && !catchingUp {
 					success = true
+				} else if ae.pindex == n.pindex {
```
Comment (mprimi) [inline server/raft.go:3408]: I don't get this bit. 
If our previous index already matches, what are we truncating?

File: server/raft.go [hunk 6]
```diff
@@ -3383,35 +3383,37 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
...
+					// Check if only our terms do not match here.
+					// Make sure pterms match and we take on the leader's.
+					// This prevents constant spinning.
+					n.truncateWAL(ae.pterm, ae.pindex)
 				} else {
 					n.resetWAL()
 				}
-			} else if eae.term != ae.term {
+			} else {
```
Comment (mprimi) [inline server/raft.go:3416]: This `else` is now broader, for example if `commit==0` and wal is empty, it will enter here. Is that expected?

File: server/raft_helpers_test.go [hunk 7]
```diff
@@ -317,3 +317,34 @@ func (rg smGroup) waitOnTotal(t *testing.T, expected int64) {
...
+	require_NoError(t, err)
+
+	cleanup := func() {
+		c.shutdown()
+	}
+	return n, cleanup
+}
+
+// Encode an AppendEntry.
+// An AppendEntry is encoded into a buffer and that's stored into the WAL.
```
Comment (mprimi) [inline server/raft_helpers_test.go:342]: I don't get the reference to the WAL here.
Looks to me like this is pure side-effect helper: attach a buffer to the AE which contains a serialized version of the AE itself.

**Review by mprimi — COMMENTED** [review]
I'm a little confused about some of these changes, but if you and Neil are confident, so am i.

It would be easier to review if each commit was atomic / self-contained and the commit message explained what the intent / reason was. 

You have some comments in GH which is convenient, but if you look at this commit in a month will you find them?
Unlike GH comments, commit messages become part of history along with the change.

File: server/raft.go (hunk 5 above)
Comment (MauriceVanVeen) [reply server/raft.go:3408]: Realistically, only updating the `pterm` when the passed `index == n.pindex`.

We could possibly also just set `n.pterm = ae.pterm` directly instead, but not sure why it was previously chosen to be a call to `truncateWAL` (so am just doing the same thing as what was done previously).

File: server/raft_helpers_test.go (hunk 7 above)
Comment (MauriceVanVeen) [reply server/raft_helpers_test.go:342]: The AppendEntry needs to be encoded into a buffer, and that buffer is what is stored into the WAL (outside of this encoding part). Without encoding we will not be able to store it.

Could maybe rephrase it to make that clearer?

File: server/raft.go (hunk 4 above)
Comment (MauriceVanVeen) [reply server/raft.go:3389]: This is how the debug statement was prior to me removing the equal check in the previous PR.

Have rephrased it to mention it can also be equal.

**Comment by derekcollison** [comment]
@mprimi you good with this PR?

**Comment by mprimi** [comment]
@derekcollison yes, we reviewed this morning and i am onboard.
